src/main/scala/devices/tilelink/CLINT.scala

```diff
@@ -12,108 +12,46 @@ import freechips.rocketchip.subsystem._
-{
-  def address = AddressSet(baseAddress, CLINTConsts.size-1)
-}
-
-case object CLINTKey extends Field[Option[CLINTParams]](None)
-
-case class CLINTAttachParams(
-  slaveWhere: TLBusWrapperLocation = CBUS
+case class CLINTParams(
+  mtimer: Option[MTIMERParams]  = Some(MTIMERParams(isCompatibleWithClint=true)),
+  mswi: Option[MSWIParams]      = Some(MSWIParams())
 )
 
-case object CLINTAttachKey extends Field(CLINTAttachParams())
-
-class CLINT(params: CLINTParams, beatBytes: Int)(implicit p: Parameters) extends LazyModule
-{
-  import CLINTConsts._
-
-  // clint0 => at most 4095 devices
-  val device = new SimpleDevice("clint", Seq("riscv,clint0")) {
-    override val alwaysExtended = true
-  }
-
-  val node: TLRegisterNode = TLRegisterNode(
-    address   = Seq(params.address),
-    device    = device,
-    beatBytes = beatBytes)
-
-  val intnode : IntNexusNode = IntNexusNode(
-    sourceFn = { _ => IntSourcePortParameters(Seq(IntSourceParameters(ints, Seq(Resource(device, "int"))))) },
-    sinkFn   = { _ => IntSinkPortParameters(Seq(IntSinkParameters())) },
-    outputRequiresInput = false)
-
-  lazy val module = new Impl
-  class Impl extends LazyModuleImp(this) {
-    Annotated.params(this, params)
-    require (intnode.edges.in.size == 0, "CLINT only produces interrupts; it does not accept them")
-
-    val io = IO(new Bundle {
-      val rtcTick = Input(Bool())
-    })
-
-    val time = RegInit(0.U(timeWidth.W))
-    when (io.rtcTick) { time := time + 1.U }
-
-    val nTiles = intnode.out.size
-    val timecmp = Seq.fill(nTiles) { Reg(UInt(timeWidth.W)) }
-    val ipi = Seq.fill(nTiles) { RegInit(0.U(1.W)) }
+case object CLINTKey extends Field[Option[CLINTParams]](None)
 
-    val (intnode_out, _) = intnode.out.unzip
-    intnode_out.zipWithIndex.foreach { case (int, i) =>
-      int(0) := ShiftRegister(ipi(i)(0), params.intStages) // msip
-      int(1) := ShiftRegister(time.asUInt >= timecmp(i).asUInt, params.intStages) // mtip
+// Actually, it's an aclint 
+trait CanHavePeripheryCLINT { this: BaseSubsystem =>
+  val aclintOpt = p(CLINTKey).map { aclintParams =>
+    val mtimerOpt = aclintParams.mtimer.map { params =>
+      val tlbus = locateTLBusWrapper(p(MTIMERAttachKey).slaveWhere)
+      val mtimer = LazyModule(new MTIMER(params, cbus.beatBytes))
+      mtimer.mtimerNode := tlbus.coupleTo("mtimer") { TLFragmenter(tlbus) := _ }
+      mtimer.mtimeSeparateNode match {
+        case Some(node) => node := tlbus.coupleTo("separateMtime") { TLFragmenter(tlbus) := _ }
+        case None => None
+      }
+
+      InModuleBody {
+        mtimer.module.clock := tlbus.module.clock
+        mtimer.module.reset := tlbus.module.reset
+      }
+
+      mtimer
     }
 
-    /* 0000 msip hart 0
-     * 0004 msip hart 1
-     * 4000 mtimecmp hart 0 lo
-     * 4004 mtimecmp hart 0 hi
-     * 4008 mtimecmp hart 1 lo
-     * 400c mtimecmp hart 1 hi
-     * bff8 mtime lo
-     * bffc mtime hi
-     */
```

Comment:
Preserving this comment somewhere would be very helpful.. with ACLINT, it becomes a bit more difficult to understand the address map

Reply:
Ok, I will write some comments.